File: app/src/main/java/com/nikola/jakshic/dagger/repository/HeroRepository.java
package com.nikola.jakshic.dagger.repository;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.arch.paging.DataSource;
import android.arch.paging.LivePagedListBuilder;
import android.arch.paging.PagedList;

import com.nikola.jakshic.dagger.AppExecutors;
import com.nikola.jakshic.dagger.Status;
import com.nikola.jakshic.dagger.data.local.HeroDao;
import com.nikola.jakshic.dagger.data.remote.OpenDotaService;
import com.nikola.jakshic.dagger.model.Hero;

import java.util.List;

import javax.inject.Inject;
import javax.inject.Singleton;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

@Singleton
public class HeroRepository {

    private OpenDotaService service;
    private HeroDao heroDao;
    private AppExecutors executor;
    private PagedList.Config config;

    @Inject
    public HeroRepository(AppExecutors executor, OpenDotaService service, HeroDao heroDao) {
        this.service = service;
        this.executor = executor;
        this.heroDao = heroDao;
        // TODO this should be provided by dagger
        config = new PagedList.Config.Builder()
                .setPrefetchDistance(15)
                .setInitialLoadSizeHint(80)
                .setPageSize(40)
                .setEnablePlaceholders(false).build();
    }

    // PagedList is not mutable, so we can not sort the items, but
    // instead we need to request new sorted data from the database
    public LiveData<PagedList<Hero>> fetchByGames(long id) {
        DataSource.Factory<Integer, Hero> factory = heroDao.getHeroesByGames(id);
        return new LivePagedListBuilder<>(factory, config).build();
    }

    // PagedList is not mutable, so we can not sort the items, but
    // instead we need to request new sorted data from the database
    public LiveData<PagedList<Hero>> fetchByWinrate(long id) {
        DataSource.Factory<Integer, Hero> factory = heroDao.getHeroesByWinrate(id);
        return new LivePagedListBuilder<>(factory, config).build();
    }

    // PagedList is not mutable, so we can not sort the items, but
    // instead we need to request new sorted data from the database
    public LiveData<PagedList<Hero>> fetchByWins(long id) {
        DataSource.Factory<Integer, Hero> factory = heroDao.getHeroesByWins(id);
        return new LivePagedListBuilder<>(factory, config).build();
    }

    // PagedList is not mutable, so we can not sort the items, but
    // instead we need to request new sorted data from the database
    public LiveData<PagedList<Hero>> fetchByLosses(long id) {
        DataSource.Factory<Integer, Hero> factory = heroDao.getHeroesByLosses(id);
        return new LivePagedListBuilder<>(factory, config).build();
    }

    /**
     * Fetches data from the network and inserts it into database
     *
     * @param status observed by the UI, so it can show proper progress bar
     * @param id player id
     */
    public void fetchHeroes(MutableLiveData<Status> status, long id) {
        status.setValue(Status.LOADING);
        service.getHeroes(id).enqueue(new Callback<List<Hero>>() {
            @Override
            public void onResponse(Call<List<Hero>> call, Response<List<Hero>> response) {
                if (response.body() != null && response.isSuccessful()) {
                    executor.diskIO().execute(() -> {
                        for (Hero item : response.body())

                            // Json response doesn't contain player id
                            // so we need to set it manually
                            item.setAccountId(id);

                        heroDao.insertHeroes(response.body());
                        status.postValue(Status.SUCCESS);
                    });
                } else
                    status.setValue(Status.ERROR);
            }

            @Override
            public void onFailure(Call<List<Hero>> call, Throwable t) {
                status.setValue(Status.ERROR);
            }
        });
    }
}

File: app/src/main/java/com/nikola/jakshic/dagger/repository/HeroRepository.kt
package com.nikola.jakshic.dagger.repository

import androidx.lifecycle.LiveData
import com.nikola.jakshic.dagger.data.local.HeroDao
import com.nikola.jakshic.dagger.data.remote.OpenDotaService
import kotlinx.coroutines.experimental.Dispatchers
import kotlinx.coroutines.experimental.withContext
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class HeroRepository @Inject constructor(
        private val dao: HeroDao,
        private val service: OpenDotaService) {

    /**
     * Constructs the [LiveData] which emits every time
     * the requested data in the database has changed
     */
    fun getHeroesLiveDataByGames(id: Long) = dao.getHeroesByGames(id)

    /**
     * Constructs the [LiveData] which emits every time
     * the requested data in the database has changed
     */
    fun getHeroesLiveDataByWinrate(id: Long) = dao.getHeroesByWinrate(id)

    /**
     * Constructs the [LiveData] which emits every time
     * the requested data in the database has changed
     */
    fun getHeroesLiveDataByWins(id: Long) = dao.getHeroesByWins(id)

    /**
     * Constructs the [LiveData] which emits every time
     * the requested data in the database has changed
     */
    fun getHeroesLiveDataByLosses(id: Long) = dao.getHeroesByLosses(id)

    /**
     * Fetches the heroes from the network and inserts them into database.
     *
     * Whenever the database is updated, the observers of [LiveData]
     * returned by [getHeroesLiveDataByGames], [getHeroesLiveDataByWins],
     * [getHeroesLiveDataByLosses] and [getHeroesLiveDataByWinrate] are notified.
     */
    suspend fun fetchHeroes(id: Long, onSuccess: () -> Unit, onError: () -> Unit) {
        try {
            val heroes = service.getHeroes(id).await()
            withContext(Dispatchers.IO) {
                heroes.map {
                    it.accountId = id   // response from the network doesn't contain any information
                    it          // about who played this heroes, so we need to set this manually
                }
                dao.insertHeroes(heroes)
            }
            onSuccess()
        } catch (e: Exception) {
            onError()
        }
    }
}